HPCC-33155 Revisit Parquet Test Suite

jackdelv opened the pull request:
Change the plugin to throw an error if a file is not found when opening a file for reading

Change output filename of partitioned file parts. part_00_0.parquet -> part_0_of_table_0_from_worker_0.parquet
These files aren't meant to be read directly, but I think the new format makes more sense when trying to debug.

Update tests so that they are all being run and are able to use external files in testing/regress/download

Remove parquetString test and key file and move testing to parquetTypes.ecl

Update test cases in parquetTypes.ecl

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jackdelv committed:
HPCC-33155 Revisit Parquet Test Suite

Signed-off-by: Jack Del Vecchio <jack.delvecchio@lexisnexisrisk.com>

dcamper commented on testing/regress/ecl/parquetPartition.ecl:
```diff
@@ -40,7 +43,51 @@ rowSize := 1024;  // Increased buffer size
+ * ├── district=1
+ * │   ├── firstname=Alice
+ * │   │   └── city=New%20York
+ * │   │       └── part_0_of_table_0_from_worker_0.parquet
```
These part... names are files? If so, they are all the same for every value combination. Is that correct?

jackdelv replied:
Yes, that is correct.

dcamper reviewed (commented):
Just one question about the contents of a comment block.

dcamper reviewed (approved):
Looks good to me.

ghalliday commented on testing/regress/ecl/parquetCorrupt.ecl:
```diff
@@ -28,12 +27,8 @@ RECORDDEF := RECORD
-                     DATASET([{0, 'Corrupt Parquet File', ''}], RECORDDEF),
-                     CORRUPT_PARQUET);
-
-OUTPUT(CORRUPT_RESULT);
```
Without the output this workunit will not do anything.

jackdelv replied:
Fixed.

ghalliday commented on testing/regress/ecl/parquetEmpty.ecl:
```diff
@@ -31,6 +36,3 @@ EMPTY_PARQUET := DATASET([], RECORDDEF);
 
 read_data := ParquetIO.Read(RECORDDEF, filePath);
-
-OUTPUT(read_data);
```
With this deleted it will not read the file.

jackdelv replied:
Fixed.

ghalliday reviewed (commented):
See comments/questions.

jackdelv committed:
Add a sort to vector of FileReaders
- createDirectoryIterator is used to list all matching Parquet files and the order is not guarenteed
- Fix parquetEmpty.ecl and parquetCorrupt.ecl by adding the OUTPUT statements back

jackdelv committed:
Fix partitioned file reading code

dcamper reviewed (approved):
Approved.

ghalliday commented on plugins/parquet/parquetembed.cpp:
```diff
@@ -430,7 +441,7 @@ arrow::Result<std::shared_ptr<arrow::Table>> ParquetReader::queryRows()
     {
         // Start by getting the number of rows in the first group and checking if it includes this workers startRow
         __int64 offset = (*rbatchItr)->get()->num_rows();
-        while (offset < startRow)
```
This looks like a bug fix - is it?

jackdelv replied:
Yes, I think the changes to the openReadFile would be considered a bug fix as well. Should they be pulled out into separate PRs?

ghalliday reviewed (commented):
Two questions.

jackdelv commented on plugins/parquet/parquetembed.cpp:
```diff
@@ -381,7 +392,7 @@ arrow::Status ParquetReader::processReadFile()
         PARQUET_ASSIGN_OR_THROW(auto datasetRows, scanner->CountRows());
         // Divide the work among any number of workers
-        divide_row_groups(activityCtx, datasetRows, totalRowCount, startRowGroup);
+        divide_row_groups(activityCtx, datasetRows, totalRowCount, startRow);
```
Yes, that was deliberate. startRow is used for arrow partitioned datasets because the API returns rows rather than groups of rows. This was an error causing multi node thor workers to always start reading from the first row.

ghalliday commented:
@jackdelv how important are the bug fixes?  Should they be targeting an earlier build than 9.10.x?
I don't mind merging as-is if we consider parquet as something that users are unlikely to be using at the moment.  If we consider it possibly in use, then they should be separated out.

AttilaVamos commented:
@jackdelv I raised a JIRA, the HPCC-33278 related to Test Suite.
